File: leetcode/173-Binary-Search-Tree-Iterator/BinSearchTreeIterator_001.java
/**
 * Definition for binary tree
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */

import java.util.Stack;

public class BSTIterator {
    private Stack<TreeNode> stack;
    private TreeNode curt;

    public BSTIterator(TreeNode root) {
        stack = new Stack<TreeNode>();
        curt = root;
    }

    /** @return whether we have a next smallest number */
    public boolean hasNext() {
        return (curt != null) || (!stack.empty());
    }

    /** @return the next smallest number */
    public int next() {
        while (curt != null) {
            stack.push(curt);
            curt = curt.left;
        }
        TreeNode res = stack.pop();
        curt = res.right;

        return res.val;
    }
}

/**
 * Your BSTIterator will be called like this:
 * BSTIterator i = new BSTIterator(root);
 * while (i.hasNext()) v[f()] = i.next();
 */

File: leetcode/173-Binary-Search-Tree-Iterator/BinSearchTreeIterator_001.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class BSTIterator:
    # @param root, a binary search tree's root node
    def __init__(self, root):
        self.curt = root
        self.stack = []

    # @return a boolean, whether we have a next smallest number
    def hasNext(self):
        return self.curt or len(self.stack) != 0

    # @return an integer, the next smallest number
    def next(self):
        curt = self.curt
        stack = self.stack
        while curt:
            stack.append(curt)
            curt = curt.left
        curt = stack.pop()
        self.curt = curt.right
        return curt.val

# Your BSTIterator will be called like this:
# i, v = BSTIterator(root), []
# while i.hasNext(): v.append(i.next())
